### PR TITLE
feat(atomWithLocation): override `replace` for specific navigations

### PR DESCRIPTION
**Pull request by Flirre**
Sometimes you want to be able to both push and replace when navigating in the same file, without having to use `useAtom` twice creating two copies of the setLocation method.

This PR allows you to override your "replace" option for specific calls to the `setLocation` call.

### CODE REVIEW

File: src/atomWithLocation.ts [hunk 1]
```diff
@@ -80,7 +80,13 @@ export function atomWithLocation<T>(options?: Options<T>) {
...
-    (get, set, arg: SetStateAction<T>) => {
+    (
+      get,
+      set,
+      arg: SetStateAction<T>,
+      atomOptions: Pick<Options<T>, 'replace'> = {},
+    ) => {
+      options = { ...options, ...atomOptions };
       set(baseAtom, arg);
       appL(get(baseAtom), options);
```
Comment (dai-shi) [inline src/atomWithLocation.ts:91]: We shouldn't override it for long, just for this call.
```suggestion
      set(baseAtom, arg);
      appL(get(baseAtom), { ...options, ...atomOptions });
```

**Review by dai-shi — CHANGES_REQUESTED** [review]
looks good otherwise.

File: __tests__/atomWithLocation_spec.tsx [hunk 2]
```diff
@@ -114,6 +114,75 @@ describe('atomWithLocation', () => {
...
+          </button>
+          <button
+            type="button"
+            onClick={() =>
+              setLocation(
+                {
+                  pathname: '/2',
+                },
+                // @ts-expect-error: TODO: How can we fix this error? Expected 1 arguments, but got 2
+                { replace: true },
```
Comment (Flirre) [inline __tests__/atomWithLocation_spec.tsx:140]: Not sure how to update type to allow this to be passed?

File: __tests__/atomWithLocation_spec.tsx [hunk 3]
```diff
@@ -114,6 +114,75 @@ describe('atomWithLocation', () => {
...
+      </StrictMode>,
+    );
+
+    const previousTestHistoryLength = 3;
+
+    await findByText('current pathname in atomWithLocation: /');
+    expect(window.location.pathname).toEqual('/');
+    expect(window.history.length).toEqual(previousTestHistoryLength);
+
+    await userEvent.click(getByText('button1')); // TODO: Why doesn't it increment the history length?
```
Comment (Flirre) [inline __tests__/atomWithLocation_spec.tsx:162]: For some unknown reason history length doesn't increment when clicking it once here.

But it increments if you click it two times.

File: __tests__/atomWithLocation_spec.tsx [hunk 4]
```diff
@@ -114,6 +114,75 @@ describe('atomWithLocation', () => {
...
+
+    await findByText('current pathname in atomWithLocation: /1');
+    expect(window.location.pathname).toEqual('/1');
+    expect(window.history.length).toEqual(previousTestHistoryLength);
+
+    await userEvent.click(getByText('button2'));
```
Comment (Flirre) [inline __tests__/atomWithLocation_spec.tsx:168]: This correctly does not increment history length, for one or several clicks.

File: __tests__/atomWithLocation_spec.tsx (hunk 2 above)
Comment (dai-shi) [reply __tests__/atomWithLocation_spec.tsx:140]: `PrimitiveAtom<Location>;` and `PrimitiveAtom<T>;` should be something like this:
```ts
type Option = { replace: boolean };

WritableAtom<Location, [SetStateAction<Location>, Option?], void>
```

File: src/atomWithLocation.ts [hunk 5]
```diff
@@ -50,13 +50,15 @@ type Options<T> = {
 type RequiredOptions<T> = Omit<Options<T>, 'getLocation' | 'applyLocation'> &
   Required<Pick<Options<T>, 'getLocation' | 'applyLocation'>>;
 
+type Option = { replace: boolean };
```
Comment (dai-shi) [inline src/atomWithLocation.ts:53]:
```suggestion
type AtomOptions = Pick<Options<T>, 'replace'>;
```

File: src/atomWithLocation.ts [hunk 6]
```diff
@@ -50,13 +50,15 @@ type Options<T> = {
...
+
 export function atomWithLocation(
   options?: Options<Location>,
-): PrimitiveAtom<Location>;
+): WritableAtom<Location, [SetStateAction<Location>, Option?], void>;
```
Comment (dai-shi) [inline src/atomWithLocation.ts:57]:
```suggestion
): WritableAtom<Location, [SetStateAction<Location>, AtomOptions?], void>;
```

File: src/atomWithLocation.ts [hunk 7]
```diff
@@ -50,13 +50,15 @@ type Options<T> = {
...
 
 export function atomWithLocation<T>(
   options: RequiredOptions<T>,
-): PrimitiveAtom<T>;
+): WritableAtom<T, [SetStateAction<T>, Option?], void>;
```
Comment (dai-shi) [inline src/atomWithLocation.ts:61]:
```suggestion
): WritableAtom<T, [SetStateAction<T>, AtomOptions?], void>;
```

File: src/atomWithLocation.ts [hunk 8]
```diff
@@ -80,9 +82,14 @@ export function atomWithLocation<T>(options?: Options<T>) {
   };
   const derivedAtom = atom(
     (get) => get(baseAtom),
-    (get, set, arg: SetStateAction<T>) => {
+    (
+      get,
+      set,
+      arg: SetStateAction<T>,
+      atomOptions: Pick<Options<T>, 'replace'> = {},
```
Comment (dai-shi) [inline src/atomWithLocation.ts:89]:
```suggestion
      atomOptions: AtomOptions = {},
```

**Review by dai-shi — COMMENTED** [review]
How about this?

File: src/atomWithLocation.ts (hunk 7 above)
Comment (Flirre) [reply src/atomWithLocation.ts]: Done in https://github.com/jotaijs/jotai-location/pull/28/commits/df353647ce1470ee858b71a1e07320efe33759d4

File: src/atomWithLocation.ts (hunk 6 above)
Comment (Flirre) [reply src/atomWithLocation.ts]: Done in https://github.com/jotaijs/jotai-location/pull/28/commits/df353647ce1470ee858b71a1e07320efe33759d4

File: src/atomWithLocation.ts (hunk 5 above)
Comment (Flirre) [reply src/atomWithLocation.ts]: Done in https://github.com/jotaijs/jotai-location/pull/28/commits/df353647ce1470ee858b71a1e07320efe33759d4

**Review by dai-shi — APPROVED** [review]
Thanks for your contribution!

**Comment by dai-shi** [comment]
Published: https://www.npmjs.com/package/jotai-location/v/0.5.3

**Pull request merged**